### PR TITLE
[FSDP2] Error on shared params in separate FSDP groups, fix test

### PR DESCRIPTION
**Pull request by wconstab**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #176257
* #176225

Address PR feedback: the previous test worked around invalid
fully_shard wrapping by manually re-tying weights after separate
fully_shard calls.

Instead:
- Add error detection in _init_sharded_param: when a non-DTensor
  parameter has already been sharded by another FSDP group (via the
  _fsdp_managed marker), raise ValueError directing users to use
  fully_shard([module_a, module_b]) to place shared/tied parameters
  in the same FSDP group.
- Update the uneven shard test to use the correct pattern:
  fully_shard([model.tok_embeddings, model.output]).
- Add a test verifying that separate fully_shard calls on modules with
  shared parameters raises the expected error.

Authored with Claude.

### CODE REVIEW

**Comment by wconstab** [comment]
squashed into previous

File: test/distributed/_composable/fsdp/test_fully_shard_training.py [hunk 1]
```diff
@@ -1086,16 +1086,39 @@ def forward(self, ids):
 
         model.tok_embeddings.unshard()
         final_weight = model.tok_embeddings.weight
-        # Every row should be updated after training. With the bug, the last
-        # rank's shard (the rows that require padding) would remain at its
-        # initial values because the all-gather reads from a stale buffer.
+        # Every row should be updated after training, including on the last
+        # rank whose shard requires padding for uneven sharding.
         for row in range(vocab_size):
```
Comment (weifengpy) [inline test/distributed/_composable/fsdp/test_fully_shard_training.py:1091]: self.assertEqual(initial_weight, final_weight) might be enough

File: torch/distributed/fsdp/_fully_shard/_fsdp_param.py [hunk 2]
```diff
@@ -287,6 +287,16 @@ def _init_sharded_param(
...
+        if not self.is_dtensor and getattr(param, "_fsdp_managed", False):
+            raise ValueError(
+                f"Parameter '{self._module_info.param_name}' is shared with a "
+                f"parameter already managed by another FSDP group. "
+                f"For shared/tied parameters, use "
+                f"fully_shard([module_a, module_b]) to place them in the same "
+                f"FSDP group."
+            )
+        if not self.is_dtensor:
+            param._fsdp_managed = True
```
Comment (weifengpy) [inline torch/distributed/fsdp/_fully_shard/_fsdp_param.py:299]: I am checking if there is a better way than duck typing `_fsdp_managed`. this is not so linter friendly

Comment (wconstab) [reply torch/distributed/fsdp/_fully_shard/_fsdp_param.py:299]: i didn't read it yet.. i can take a real look tmrw and try to make it nice if you think the fix is mostly good.

File: test/distributed/_composable/fsdp/test_fully_shard_training.py (hunk 1 above)
Comment (weifengpy) [reply test/distributed/_composable/fsdp/test_fully_shard_training.py:1091]: the bug report is about staled weights (no optimizer updates) on last rank for uneven sharding. maybe do a few opt.step here with uneven weight to verify?

Comment (wconstab) [reply test/distributed/_composable/fsdp/test_fully_shard_training.py:1091]: 1. On assertEqual vs the per-row loop:                                                                              
                                                                                                                      
  No, self.assertEqual(initial_weight, final_weight) would assert the tensors are equal — the opposite of what you    
  want. And even if you meant self.assertNotEqual, that would only confirm that at least one element differs. The     
  whole point of this test is to verify that every row gets updated, including rows on the last rank whose shard
  requires padding for uneven sharding. That's the bug this test covers. A coarse assertNotEqual would pass even if
  the padded rows on the last rank were silently zeroed out / never updated — exactly the failure mode being tested.
  The per-row loop with the diagnostic message (f"Row {row} was not updated") is the right approach here.
-Claude

File: torch/distributed/fsdp/_fully_shard/_fsdp_param.py (hunk 2 above)
Comment (wconstab) [reply torch/distributed/fsdp/_fully_shard/_fsdp_param.py:299]: 2. On _fsdp_managed duck typing:

  You're right that param._fsdp_managed = True / getattr(param, "_fsdp_managed", False) is duck typing on nn.Parameter
   (a tensor). It's only used in two places in _fsdp_param.py (lines 290 and 299), and it's never cleaned up.

  A cleaner alternative would be a module-level WeakSet that tracks managed params by identity:

  _fsdp_managed_params: weakref.WeakSet = weakref.WeakSet()

  Then replace:
  - param._fsdp_managed = True → _fsdp_managed_params.add(param)
  - getattr(param, "_fsdp_managed", False) → param in _fsdp_managed_params

  This avoids monkey-patching the tensor, is linter friendly, and the WeakSet automatically cleans up when the
  parameter is garbage collected. It's also a pattern already used elsewhere in PyTorch for tracking tensor state
  without modifying the tensor itself.

  Note that _is_fsdp_managed_module (on modules, line 224 of _fully_shard.py) is a separate concern — that one is read
   by Dynamo (builder.py:2007) so changing it has a wider blast radius. But for _fsdp_managed on params, it's
  self-contained in _fsdp_param.py and straightforward to convert.

  Want me to make the WeakSet change?

File: test/distributed/_composable/fsdp/test_fully_shard_training.py (hunk 1 above)
Comment (weifengpy) [reply test/distributed/_composable/fsdp/test_fully_shard_training.py:1091]: you're right. I didn't realized this assersion is after training loop. I mark this as resolved. no change needed

File: torch/distributed/fsdp/_fully_shard/_fsdp_param.py (hunk 2 above)
Comment (weifengpy) [reply torch/distributed/fsdp/_fully_shard/_fsdp_param.py:299]: Maybe use existing hierachy to detect duplicate parameters: FSDPModule -> FSDPState -> FSDPParamGroup -> FSDPParam. It means duplicates detection in lazy_init

```
seen = set()
  for state in self._state_ctx.all_states:
      if group := state._fsdp_param_group:
          for fsdp_param in group.fsdp_params:
              param = fsdp_param._orig_param  # or however you get the original
              if id(param) in seen:
                  raise ValueError(...)
              seen.add(id(param))
```
